onelogin_rules/onelogin_unusual_login.py

```diff
@@ -0,0 +1,60 @@
+import json
+import requests
+from panther_oss_helpers import get_string_set, put_string_set
+
+
+def rule(event):
+    # Pre-filter to save compute time where possible. event_type_id = 5 is login events.
+    if event.get('event_type_id') != 5 or event.get('ipaddr') is None:
+        return False
+
+    # Lookup geo-ip data via API call
+    url = 'https://ipinfo.io/' + event['ipaddr'] + '/geo'
+
+    # Skip API call if this is a unit test
+    if __name__ == 'PolicyApiTestingPolicy':
+        resp = lambda: None
+        setattr(resp, 'status_code', 200)
+        setattr(resp, 'text', event['api_data'])
+    else:
+        resp = requests.get(url)
+
+    if resp.status_code != 200:
+        # Could raise an exception here for ops team to look into
+        return False
+    login_info = json.loads(resp.text)
+    login_tuple = login_info.get('region', '<REGION>') + ":" + login_info.get(
+        'city', '<CITY>')
+
+    # Lookup & store persistent data
+    event_key = get_key(event)
+    last_login_info = get_string_set(event_key)
+    if not last_login_info:
+        # Store this as the first login if we've never seen this user login before
+        put_string_set(event_key, [json.dumps({login_tuple: 1})])
+        return False
+    last_login_info = json.loads(last_login_info.pop())
+
+    last_login_info[login_tuple] = last_login_info.get(login_tuple, 0) + 1
+    put_string_set(event_key, [json.dumps(last_login_info)])
+
+    tuple_count = last_login_info[login_tuple]
+    higher_tuples = 0
+    for tcount in last_login_info.values():
+        if tcount > tuple_count:
+            higher_tuples += 1
+        if higher_tuples == 3:
+            return True
```

Comment:
So what does this mean? We alert if the number of "unusual" logins is greater than a threshold? How do we determine what's unusual, or is that done by the API?

Reply:
I added a comment here as well explaining the logic a bit. We could probably do some fancy python list comprehension to make this a one liner if we want, but that probably wouldn't make the logic anymore clear.